framework/sandstone_p.h

```diff
@@ -320,6 +324,67 @@ class LoggingStream
     friend LoggingStream logging_user_messages_stream(int thread_num, int level);
 };
 
+class DeviceSchedule {
+public:
+    virtual int get_next_cpu() = 0;
+    virtual ~DeviceSchedule() = default;
+};
+
+class QueueDeviceSchedule : public DeviceSchedule {
+public:
+    QueueDeviceSchedule(struct cpu_info *cpu_info, int q_size)
+    {
+        q_idx.store(0, std::memory_order_relaxed);
+        queue.reserve(q_size);
+
+        // Populate queue
+        for (int i=0; i<q_size; i++)
+            queue.push_back(cpu_info[i].cpu_number);
```

Comment:
We cannot assume that the queue is the same size as the number of available CPUs. So while this can be populated with the first `q_size` ones. I would take a different approach and have 1. a queue of size of available cpus (which is >= `q_size`) and 2. track current index (as class member). Here, populate the queue (and, optionally, shuffle) when index reaches `q_size`, shuffle the entire queue and reset index to 0.